clippy_lints/src/any_coerce.rs

```diff
@@ -0,0 +1,213 @@
+use crate::rustc::ty::adjustment::Adjust;
+use crate::rustc::ty::{self, ToPolyTraitRef, Ty};
+use crate::rustc::{declare_tool_lint, lint_array};
+use crate::syntax_pos::symbol::Ident;
+use crate::utils::{match_def_path, paths, span_lint_and_then};
+use if_chain::if_chain;
+use std::collections::VecDeque;
+
+/// **What it does:** Checks for coercing something that already contains a
+/// `dyn Any` to `dyn Any` itself.
+///
+/// **Why is this bad?** It's probably a mistake.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// let box_foo: Box<Foo> = Box::new(Foo);
+/// let mut box_any: Box<dyn Any> = box_foo;
+/// let bad: &mut dyn Any = &mut box_any;
+/// // you probably meant
+/// let ok: &mut dyn Any = &mut *box_any;
+/// ```
+declare_clippy_lint! {
+    pub WRONG_ANY_COERCE,
+    correctness,
+    "coercing a type already containing `dyn Any` to `dyn Any` itself"
+}
+
+#[derive(Default)]
+pub struct WrongAnyCoerce;
+
+impl LintPass for WrongAnyCoerce {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(WRONG_ANY_COERCE)
+    }
+}
+
+struct LintData<'tcx> {
+    coerced_to_any: Ty<'tcx>,
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for WrongAnyCoerce {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        let adjustments = cx.tables.expr_adjustments(expr);
+        for (i, adj) in adjustments.iter().enumerate() {
+            if let Adjust::Unsize = adj.kind {
+                let src_ty = if i == 0 {
+                    cx.tables.expr_ty(expr)
+                } else {
+                    adjustments[i - 1].target
+                };
+                cx.tcx.infer_ctxt().enter(|infcx| {
+                    let opt_lint_data = check_unsize_coercion(infcx, cx.param_env, src_ty, adj.target);
+                    if let Some(lint_data) = opt_lint_data {
+                        // TODO: we might be able to suggest dereferencing in some cases
+                        let cta_str = lint_data.coerced_to_any.to_string();
+                        span_lint_and_then(
+                            cx,
+                            WRONG_ANY_COERCE,
+                            expr.span,
+                            &format!("coercing `{}` to `dyn Any`", cta_str),
+                            |db| {
+                                if !cta_str.contains("Any") {
+                                    db.note(&format!("`{}` dereferences to `dyn Any`", cta_str));
+                                }
+                            },
+                        )
+                    }
+                });
+            }
+        }
+    }
+}
+
+/// Returns whether or not this coercion should be linted
+fn check_unsize_coercion<'tcx>(
+    infcx: InferCtxt<'_, '_, 'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    src_ty: Ty<'tcx>,
+    tgt_ty: Ty<'tcx>,
+) -> Option<LintData<'tcx>> {
+    // redo the typechecking for this coercion to see if it required unsizing something to `dyn Any`
```

Comment:
What's the reason for this? Would it be possible to use https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.struct_lockstep_tails to figure out whether the `src_ty` already is `Any` at the same level that `tgt_ty` is?

Reply:
`struct_lockstep_tails` doesn't go through references and `Box<T>` ends up as `PhantomData<T>`, so we'd have to do some additional work to use that. That also wouldn't work if the type is generic (i.e. `T: Deref<Target=dyn Any>`).

Reply:
ugh, makes sense :/

Do you think we can share this function with rustc by refactoring the original rustc code to factor out this function and make it public?

Reply:
Something like `resolve_obligations_for_unsize_coercion(.., obligation_callback: impl FnMut(&PredicateObligation) -> ()) -> ..`? Then typechecking could use the callback to check for unsizing to a tuple (which apparently is feature gated) and we could check for unsizing to `dyn Any`?